File: packages/runtime-dotnet/src/Support.cs
using System.Globalization;

namespace Graft.Runtime;

/// <summary>
/// Helpers generated models call.
/// </summary>
/// <remarks>
/// In the runtime rather than emitted per SDK, because they are identical for every spec and hand-written code
/// is where this project puts quality (<c>AGENTS.md</c>). Generated decoders stay a list of narrowing
/// assignments, which is the part that genuinely differs per model.
/// </remarks>
public static class Support
{
  /// <summary>
  /// A required field was absent or the wrong type.
  /// </summary>
  /// <remarks>
  /// Returns <c>T</c> rather than void so it can be the else-branch of a conditional expression, which is what
  /// lets a generated decoder narrow in one statement per field.
  /// </remarks>
  public static T Fail<T>(string owner, string field, string expected) =>
      throw new DecodeException($"{owner}: expected {expected} for `{field}`");

  /// <summary>The object at a key, or null.</summary>
  public static object? Get(object? data, string key) =>
      data is IDictionary<string, object?> map && map.TryGetValue(key, out var value) ? value : null;

  /// <summary>
  /// Parse an RFC 3339 timestamp.
  /// </summary>
  /// <remarks>
  /// Returns null on an unparseable value rather than throwing, so the caller decides whether that is fatal —
  /// which depends on whether the field was required. <c>RoundtripKind</c> keeps an explicit offset rather
  /// than silently converting to local time, which is the default and is wrong for a wire value.
  /// </remarks>
  public static DateTimeOffset? Instant(object? value)
  {
    if (value is DateTimeOffset already)
    {
      return already;
    }

    if (value is not string text || string.IsNullOrWhiteSpace(text))
    {
      return null;
    }

    return DateTimeOffset.TryParse(
        text,
        CultureInfo.InvariantCulture,
        DateTimeStyles.RoundtripKind,
        out var parsed)
        ? parsed
        : null;
  }

  /// <summary>Decode a JSON array, keeping nulls: a JSON array is allowed to contain them.</summary>
  public static IReadOnlyList<T> List<T>(object? value, Func<object?, T> decode)
  {
    if (value is not IList<object?> items)
    {
      return Array.Empty<T>();
    }

    var result = new List<T>(items.Count);
    foreach (var item in items)
    {
      result.Add(decode(item));
    }

    return result;
  }

  /// <summary>Decode a JSON object as a map.</summary>
  public static IReadOnlyDictionary<string, T> Map<T>(object? value, Func<object?, T> decode)
  {
    // An empty map arrives as `[]` from a PHP backend, which is the artifact SPEC.md §3.1.2 names.
    if (value is IList<object?> { Count: 0 })
    {
      return new Dictionary<string, T>();
    }

    if (value is not IDictionary<string, object?> map)
    {
      return new Dictionary<string, T>();
    }

    var result = new Dictionary<string, T>(map.Count);
    foreach (var (key, item) in map)
    {
      result[key] = decode(item);
    }

    return result;
  }

  /// <summary>
  /// A paginated method's base query, plus the parameters the paginator advances.
  /// </summary>
  /// <remarks>
  /// The paginator's parameters win, because they are the ones that change per page — a base <c>offset</c>
  /// would otherwise pin every request to the first page.
  /// </remarks>
  public static IReadOnlyDictionary<string, object?> Merged(
      IReadOnlyDictionary<string, object?> baseQuery,
      IReadOnlyDictionary<string, object?> advancing)
  {
    var result = new Dictionary<string, object?>(baseQuery);
    foreach (var (key, value) in advancing)
    {
      result[key] = value;
    }

    return result;
  }
}

File: packages/runtime-java/src/main/java/graft/runtime/Support.java
package graft.runtime;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * Helpers generated models call.
 *
 * <p>In the runtime rather than emitted per SDK, because they are identical for every spec and
 * hand-written code is where this project puts quality ({@code AGENTS.md}). Generated decoders stay
 * a list of narrowing assignments, which is the part that genuinely differs per model.
 */
public final class Support {

  private Support() {}

  /**
   * A required field was absent or the wrong type.
   *
   * <p>Returns {@code <T>} rather than {@code void} so it can be used as the else-branch of a
   * conditional expression, which is what lets a generated decoder narrow in one statement per
   * field.
   */
  public static <T> T fail(String owner, String field, String expected) {
    throw new DecodeException(owner + ": expected " + expected + " for `" + field + "`");
  }

  /**
   * Parse an RFC 3339 timestamp, tolerating an offset.
   *
   * <p>Returns null on an unparseable value rather than throwing, so the caller decides whether
   * that is fatal — which depends on whether the field was required. {@code Instant.parse} rejects
   * an offset other than {@code Z}, which real APIs send, so {@code OffsetDateTime} is tried
   * second.
   */
  public static Instant instant(Object value) {
    if (value instanceof Instant already) {
      return already;
    }
    if (!(value instanceof String text) || text.isBlank()) {
      return null;
    }
    try {
      return Instant.parse(text);
    } catch (java.time.format.DateTimeParseException ignored) {
      try {
        return OffsetDateTime.parse(text).toInstant();
      } catch (java.time.format.DateTimeParseException alsoIgnored) {
        return null;
      }
    }
  }

  /** Decode a JSON array, dropping nothing: a null element stays null. */
  public static <T> List<T> list(Object value, Function<Object, T> decode) {
    if (!(value instanceof List<?> items)) {
      return List.of();
    }
    List<T> out = new ArrayList<>(items.size());
    for (Object item : items) {
      out.add(decode.apply(item));
    }
    // `Collections.unmodifiableList`, not `List.copyOf`: the latter rejects null elements, and a
    // JSON array
    // is allowed to contain them.
    return java.util.Collections.unmodifiableList(out);
  }

  /** Decode a JSON object as a map. */
  public static <T> Map<String, T> mapOf(Object value, Function<Object, T> decode) {
    if (value instanceof List<?> empty && empty.isEmpty()) {
      // An empty map arrives as `[]` from a PHP backend, which is the artifact SPEC.md §3.1.2
      // names.
      return Map.of();
    }
    if (!(value instanceof Map<?, ?> map)) {
      return Map.of();
    }
    LinkedHashMap<String, T> out = new LinkedHashMap<>();
    map.forEach((key, item) -> out.put(String.valueOf(key), decode.apply(item)));
    return java.util.Collections.unmodifiableMap(out);
  }

  /**
   * Encode one value for a request body: a model's tree, an enum's wire value, or the value itself.
   */
  public static Object encodeOne(Object value) {
    if (value == null) {
      return null;
    }
    if (value instanceof Query.WireValued wired) {
      return wired.wireValue();
    }
    if (value instanceof Instant instant) {
      return instant.toString();
    }
    if (value instanceof Json.JsonValue custom) {
      return custom.toJson();
    }
    return value;
  }

  /** Encode a list for a request body. */
  public static <T> List<Object> encodeList(List<T> values, Function<T, Object> encode) {
    List<Object> out = new ArrayList<>(values.size());
    for (T value : values) {
      out.add(encode.apply(value));
    }
    return out;
  }

  /**
   * A paginated method's base query, plus the parameters the paginator advances.
   *
   * <p>The paginator's parameters win, because they are the ones that change per page — a base
   * {@code offset} would otherwise pin every request to the first page. A {@code LinkedHashMap}
   * rather than {@code Map.of}, because an omitted optional parameter is a null value and {@code
   * Map.of} rejects those.
   */
  public static Map<String, Object> merged(Map<String, ?> base, Map<String, ?> advancing) {
    LinkedHashMap<String, Object> out = new LinkedHashMap<>(base);
    out.putAll(advancing);
    return out;
  }

  /** Encode a map for a request body. */
  public static <T> Map<String, Object> encodeMap(
      Map<String, T> values, Function<T, Object> encode) {
    LinkedHashMap<String, Object> out = new LinkedHashMap<>();
    values.forEach((key, value) -> out.put(key, encode.apply(value)));
    return out;
  }
}
